### ISSUE TITLE
Obtaining source span for DartObject (computeConstantValue)

### ISSUE THREAD

**Issue by matanlurey** [dart-lang/sdk#34062]
(Reference issue is https://github.com/dart-lang/angular/issues/1350)

tl;dr: AngularDart has a metadata-based API for consumers:

```dart
import 'package:angular/angular.dart';

@Component(
  selector: 'bar',
)
class Foo {}
```

Imagine we wanted to validate that `selector` was always the class name, kabob-cased. The best we can do today is present either a source span for `class Foo` or a source span for the entire `@Component` annotation - of which it can become quite large for complex components.

We'd like to be able to get a source span so we could present:

```
@Component(
  selector: 'bar',
  ^^^^^^^^^^^^^^^
  Expected 'foo' for 'class Foo', but got 'bar'
```

Unfortunately it looks like we lack the APIs in `package:analyzer` to do this. We could re-parse the file, and try to "recompute" what the source span would be, but that would likely be non-trivial and require a lot of ugly code. Any suggestions please?
